In [1]:
import findspark
findspark.init()

In [2]:
import pyspark
from pyspark.sql import SparkSession

In [3]:
spark = SparkSession.builder.getOrCreate()

23/05/26 12:36:28 WARN Utils: Your hostname, prpa resolves to a loopback address: 127.0.1.1; using 10.0.2.15 instead (on interface enp0s3)
23/05/26 12:36:28 WARN Utils: Set SPARK_LOCAL_IP if you need to bind to another address


Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).


23/05/26 12:36:40 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable


In [4]:
import json
import csv

In [8]:
enero2019 = spark.read.json('201901.json')
print ('ENERO 2019')

enero2019.count()

ENERO 2019


279223

In [52]:
enero2020 = spark.read.json('202001.json')
print ('ENERO 2020')

enero2020.count()


ENERO 2020


272875

In [15]:
enero2021 = spark.read.json('202101.json')
print ('ENERO 2021')

enero2021.count()

ENERO 2021


125642

In [6]:
enero2022 = spark.read.csv('202201.csv', sep=';', header = True)
print ('ENERO 2022')

enero2022.count()//2


ENERO 2022


244760

In [10]:
enero2023 = spark.read.csv('202301.csv', sep=';', header = True)
print ('ENERO 2023')

enero2023.count()//2

ENERO 2023


295923

In [12]:
print('Número de desplazamientos Enero 2019: ' + str(enero2019.count()))
print('Número de desplazamientos Enero 2020: ' + str(enero2020.count()))
print('Número de desplazamientos Enero 2021: ' + str(enero2021.count()))
print('Número de desplazamientos Enero 2022: ' + str(enero2022.count()//2))
print('Número de desplazamientos Enero 2023: ' + str(enero2023.count()//2))

Número de desplazamientos Enero 2019: 279223
Número de desplazamientos Enero 2020: 272875
Número de desplazamientos Enero 2021: 125642
Número de desplazamientos Enero 2022: 244760
Número de desplazamientos Enero 2023: 295923


In [20]:
unlock2019 = enero2019.groupBy('idunplug_station').count()
unlock2020 = enero2020.groupBy('idunplug_station').count()
unlock2021 = enero2021.groupBy('idunplug_station').count()
unlock2022 = enero2022.groupBy('station_unlock').count()
unlock2023 = enero2023.groupBy('station_unlock').count()


In [29]:
print ('ENERO 2019')
unlock2019.filter(unlock2019['count']>3400).show()
print ('ENERO 2020')
unlock2020.filter(unlock2020['count']>3000).show()
print ('ENERO 2021')
unlock2021.filter(unlock2021['count']>1100).show()
print ('ENERO 2022')
unlock2022.filter(unlock2022['count']>2300).show()
print ('ENERO 2023')
unlock2023.filter(unlock2023['count']>2500).show()

ENERO 2019


+----------------+-----+
|idunplug_station|count|
+----------------+-----+
|            null|47215|
|             175| 3480|
|             163| 3708|
|             129| 3513|
+----------------+-----+

ENERO 2020


+----------------+-----+
|idunplug_station|count|
+----------------+-----+
|              57| 3075|
|              43| 3674|
|              83| 3057|
|             175| 3740|
+----------------+-----+

ENERO 2021
+----------------+-----+
|idunplug_station|count|
+----------------+-----+
|              57| 1637|
|              43| 1586|
|             208| 1158|
|             175| 1182|
+----------------+-----+

ENERO 2022


+--------------+------+
|station_unlock| count|
+--------------+------+
|          null|246630|
|            43|  2846|
|           175|  2421|
|            57|  2377|
+--------------+------+

ENERO 2023


+--------------+------+
|station_unlock| count|
+--------------+------+
|          null|296714|
|            43|  3657|
|           220|  2671|
|           175|  2832|
|            57|  3752|
+--------------+------+



ESTACIONES DONDE MÁS SE COGEN BICICLETAS

ENERO 2019
- 163: Paseo de la Esperanza, 83
- 129: Calle de Jose Abascal, 33
- 175: Calle Jaime el Conquistador, 30

ENERO 2020
- 175: Calle Jaime el Conquistador, 30
- 43: Calle de Lavapiés, 36
- 57/83: Calle de Santa Cruz de Marcenado, 24 / Pio Baroja, 8B

ENERO 2021
- 57: Calle de Santa Cruz de Marcenado, 24
- 43: Calle de Lavapiés, 36
- 175/208: Calle Jaime el Conquistador, 30 / Paseo de la Castellana, S/N

ENERO 2022
- 43: Calle de Lavapiés, 36
- 175: Calle Jaime el Conquistador, 30
- 57: Calle de Santa Cruz de Marcenado, 24

ENERO 2023
- 57: Calle de Santa Cruz de Marcenado, 24
- 43: Calle de Lavapiés, 36
- 175: Calle Jaime el Conquistador, 30

En general, donde más se cogen es en la estación 175, 43, 57

In [30]:
lock2019 = enero2019.groupBy('idplug_station').count()
lock2020 = enero2020.groupBy('idplug_station').count()
lock2021 = enero2021.groupBy('idplug_station').count()
lock2022 = enero2022.groupBy('station_lock').count()
lock2023 = enero2023.groupBy('station_lock').count()

In [32]:
print ('ENERO 2019')
lock2019.filter(lock2019['count']>3300).show()
print ('ENERO 2020')
lock2020.filter(lock2020['count']>3000).show()
print ('ENERO 2021')
lock2021.filter(lock2021['count']>1100).show()
print ('ENERO 2022')
lock2022.filter(lock2022['count']>2300).show()
print ('ENERO 2023')
lock2023.filter(lock2023['count']>2500).show()

ENERO 2019


+--------------+-----+
|idplug_station|count|
+--------------+-----+
|            43| 3336|
|          null|47215|
|           175| 3608|
|           163| 3774|
|           129| 3396|
+--------------+-----+

ENERO 2020
+--------------+-----+
|idplug_station|count|
+--------------+-----+
|            57| 3057|
|            43| 3748|
|            83| 3068|
|           175| 3741|
+--------------+-----+

ENERO 2021
+--------------+-----+
|idplug_station|count|
+--------------+-----+
|           149| 1101|
|            57| 1639|
|            43| 1594|
|           208| 1179|
|           175| 1170|
+--------------+-----+

ENERO 2022


+------------+------+
|station_lock| count|
+------------+------+
|        null|247067|
|          43|  2901|
|         175|  2508|
|          57|  2418|
+------------+------+

ENERO 2023


+------------+------+
|station_lock| count|
+------------+------+
|        null|296887|
|          43|  3698|
|         220|  2848|
|         175|  2906|
|          57|  3759|
+------------+------+



ESTACIONES DONDE MÁS SE DEVUELVEN BICICLETAS

ENERO 2019
- 163: Paseo de la Esperanza, 83
- 175: Calle Jaime el Conquistador, 30
- 129/43: Calle Jose Abascal, 33 / Calle de Lavapiés, 36

ENERO 2020
- 43: Calle de Lavapiés, 36
- 175: Calle Jaime el Conquistador, 30
- 83/57: Pio Baroja, 8B / Calle de Santa Cruz de Marcenado, 24

ENERO 2021
- 57: Calle de Santa Cruz de Marcenado, 24
- 43: Calle de Lavapiés, 36
- 208/175: Paseo de la Castellana, S/N / Calle Jaime el Conquistador, 30

ENERO 2022
- 43: Calle de Lavapiés, 36
- 175: Calle Jaime el Conquistador, 30
- 57: Calle de Santa Cruz de Marcenado, 24

ENERO 2023
- 57: Calle de Santa Cruz de Marcenado, 24
- 43: Calle de Lavapiés, 36
- 175/220: Calle Jaime el Conquistador, 30 / Arroyo de las Pilillas , 4

En general, donde más se devuelven es en la estación 175, 43, 57

Se podría decir que las estaciones que más triunfan son 175, 43, 57

In [43]:
print ('ENERO 2019')
unlock2019.filter(unlock2019['count']<350).show()
print ('ENERO 2020')
unlock2020.filter(unlock2020['count']<400).show()
print ('ENERO 2021')
unlock2021.filter(unlock2021['count']<60).show()
print ('ENERO 2022')
unlock2022.filter(unlock2022['count']<20).show()
print ('ENERO 2023')
unlock2023.filter(unlock2023['count']<200).show()

ENERO 2019


+----------------+-----+
|idunplug_station|count|
+----------------+-----+
|              29|  327|
|              28|  268|
|            2008|    1|
+----------------+-----+

ENERO 2020
+----------------+-----+
|idunplug_station|count|
+----------------+-----+
|              29|  339|
|              28|  167|
+----------------+-----+

ENERO 2021
+----------------+-----+
|idunplug_station|count|
+----------------+-----+
|             257|   49|
|             265|   52|
+----------------+-----+

ENERO 2022


+--------------+-----+
|station_unlock|count|
+--------------+-----+
|           115|   11|
|           123|    9|
|           124|   12|
+--------------+-----+

ENERO 2023


+--------------+-----+
|station_unlock|count|
+--------------+-----+
|           117|  164|
|           889|    3|
|            89|    9|
+--------------+-----+



ESTACIONES DONDE MENOS SE COGEN BICICLETAS

ENERO 2019
- 2008: NO EXISTE
- 28: Calle de Alcalá, 26
- 29: Marqués de Cubas, 25

ENERO 2020
- 28: Calle de Alcalá, 26
- 29: Marqués de Cubas, 25

ENERO 2021
- 49: Ronda de Atocha, 36
- 52: Plaza de Santa Ana, 9

ENERO 2022
- 123: Guzmán el Bueno, 2
- 115/124: Plaza del Marqués de Cerralbo, S/N / Manuel Silvela, 18

ENERO 2023
- 889: NO EXISTE
- 89: Ayala, 102
- 117: Altamirano, 24
DIFERENCIAS MUY GRANDES

En general, no hay un consenso ahora mismo de dónde se cogen menos

In [46]:
print ('ENERO 2019')
lock2019.filter(lock2019['count']<350).show()
print ('ENERO 2020')
lock2020.filter(lock2020['count']<400).show()
print ('ENERO 2021')
lock2021.filter(lock2021['count']<60).show()
print ('ENERO 2022')
lock2022.filter(lock2022['count']<200).show()
print ('ENERO 2023')
lock2023.filter(lock2023['count']<200).show()

ENERO 2019


+--------------+-----+
|idplug_station|count|
+--------------+-----+
|            29|  344|
|            28|  271|
|          2008|    1|
+--------------+-----+

ENERO 2020
+--------------+-----+
|idplug_station|count|
+--------------+-----+
|            29|  332|
|            28|  170|
+--------------+-----+

ENERO 2021
+--------------+-----+
|idplug_station|count|
+--------------+-----+
|           266|   52|
|           257|   35|
|           265|   52|
|           267|   59|
|           261|   54|
+--------------+-----+

ENERO 2022


+------------+-----+
|station_lock|count|
+------------+-----+
|          28|  185|
|         257|  183|
|         266|  117|
|         225|  168|
+------------+-----+

ENERO 2023


+------------+-----+
|station_lock|count|
+------------+-----+
|          28|  180|
|         117|  160|
|         889|    4|
|          89|   12|
+------------+-----+



ESTACIONES DONDE MENOS SE DEVUELVEN BICICLETAS

ENERO 2019
- 2008: NO EXISTE
- 28: Calle de Alcalá, 26
- 29: Marqués de Cubas, 25

ENERO 2020
- 28: Calle de Alcalá, 26
- 29: Marqués de Cubas, 25

ENERO 2021
- 257: Avenida de Juan de Herrera, 1
- 266/265: NO APARECEN EN LA WEB

ENERO 2022                                                                           
- 266: NO APARECE EN LA WEB
- 225: Sierra Toledana, 33
266 SIGUE ESTANDO ABAJO EN EL RANKING PERO HA AUMENTADO CONSIDERABLEMENTE


ENERO 2023
- 889: NO EXISTE
- 89: Ayala, 102
- 117/28: Altamirano, 24 / Alcalá , 26

28 SIGUE SIN TRIUNFAR

En general, donde menos se devuelven es en la estación 28

Se podría decir que la estación menos utilizada es la 28

In [ ]:
AGERANGE, TRAVELTIME, USERTYPE 3 PRIMEROS AÑOS
FECHA MÁS UTILIZADAS DE LOS ÚLTIMOS 2

VEAMOS QUÉ AÑO SE UTILIZÓ MÁS EN ENERO

In [54]:
print ('ENERO 2019')
enero2019.groupBy('user_type').sum('travel_time').show()
print ('ENERO 2020')
enero2020.groupBy('user_type').sum('travel_time').show()
print ('ENERO 2021')
enero2021.groupBy('user_type').sum('travel_time').show()
print ('ENERO 2022')
enero2022.groupBy('user_type').sum('trip_minutes').show()
print ('ENERO 2023')
enero2023.groupBy('user_type').sum('trip_minutes').show()

ENERO 2019


+---------+----------------+
|user_type|sum(travel_time)|
+---------+----------------+
|        0|         2592405|
|     null|            null|
|        1|       165544356|
|        3|        69331926|
|        2|         2563298|
+---------+----------------+

ENERO 2020
+---------+----------------+
|user_type|sum(travel_time)|
+---------+----------------+
|        1|       211759904|
|        3|        92907755|
|        2|         1505971|
+---------+----------------+

ENERO 2021
+---------+----------------+
|user_type|sum(travel_time)|
+---------+----------------+
|        6|         1351215|
|        1|       126578813|
|        3|        27008414|
|        2|          699819|
+---------+----------------+

ENERO 2022


AnalysisException: "trip_minutes" is not a numeric column. Aggregation function can only be applied on a numeric column.

In [7]:
abril2020 = spark.read.json('202004.json')
print ('ABRIL 2020')

abril2020.count()

ABRIL 2020


10447

In [22]:
mayo2020 = spark.read.json('202005.json')
print ('MAYO 2020')

mayo2020.count()

MAYO 2020


259579

In [23]:
junio2020 = spark.read.json('202006.json')
print ('JUNIO 2020')

junio2020.count()

JUNIO 2020


210555